compiler/src/iree/compiler/Dialect/Util/Analysis/GlobalTable.h

```diff
@@ -0,0 +1,129 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_COMPILER_DIALECT_IREE_ANALYSIS_GLOBALTABLE_H_
+#define IREE_COMPILER_DIALECT_IREE_ANALYSIS_GLOBALTABLE_H_
+
+#include <functional>
+
+#include "iree/compiler/Dialect/Util/IR/UtilOps.h"
+#include "iree/compiler/Dialect/Util/IR/UtilTypes.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/Support/LLVM.h"
+
+namespace mlir::iree_compiler::IREE::Util {
+
+// An entry in a GlobalTable representing a util.global (or related) op.
+struct Global {
+  // Ordinal of the global in the parent module.
+  size_t ordinal = 0;
+
+  // Global this information relates to.
+  IREE::Util::GlobalOpInterface op;
+
+  // True if the address of the global is ever taken.
+  // This disables most optimizations here; we could use some data flow analysis
+  // to track potential operations on globals via the addresses but we don't
+  // currently have any input programs that require doing so.
+  bool isIndirect = false;
+
+  // All util.global.load ops referencing the global.
+  SmallVector<IREE::Util::GlobalLoadOpInterface> loadOps;
+  // All util.global.store ops referencing the global.
+  SmallVector<IREE::Util::GlobalStoreOpInterface> storeOps;
+  // All other operations that reference the global via an attribute.
+  SmallVector<Operation *> referencingOps;
+
+  // Returns the symbol name of the global op.
+  StringRef getName() { return op.getGlobalName().getValue(); }
+
+  // Returns true if the global is a candidate for folding.
+  bool isCandidate() { return !isIndirect && op.isGlobalPrivate(); }
```

Comment:
(this code was mostly moved from the other file, but still wanted to say how much I like this code pattern)

These helper functions for seemingly trivial composite conditions have been really helpful with enhancing readability at call sites and making changes to the conditions easier in the future. Thanks!